src/HexagonOptimize.cpp

```diff
@@ -1605,6 +1605,215 @@ class OptimizeShuffles : public IRMutator {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    // Return the ramp of first vector if all vector in exprs are continuous
+    // vectors in memory.
+    Expr are_continuous_vectors(const vector<Expr> exprs) {
+        if (exprs.size() == 0) {
+            return nullptr;
+        }
+        const Load *prev_load;
+        const Load *curr_load;
+        const Ramp *prev_ramp = nullptr;
+        const Ramp *curr_ramp = nullptr;
+        for (size_t i = 0; i < exprs.size(); i++) {
+            Expr maybe_load = calc_load(exprs[i]);
+            if (!maybe_load.defined()) {
+                return nullptr;
+            }
+            curr_load = maybe_load.as<Load>();
+            curr_ramp = get_val(curr_load->index).as<Ramp>();
+            if (!curr_ramp) {
+                return nullptr;
+            }
+            if (i > 0 && curr_load->name == prev_load->name) {
+                Expr vector_size = prev_ramp->lanes * prev_ramp->stride;
+                Expr base_diff = simplify(curr_ramp->base - prev_ramp->base - vector_size);
+                if (!is_const(base_diff, 0)) {
+                    return nullptr;
+                }
+            }
+            prev_load = curr_load;
+            prev_ramp = curr_ramp;
+        }
+        return calc_load(exprs[0]);
+    }
+
+    // Returns the load indicating vector start index. If the vector is sliced
+    // return load with shifted ramp to include slice_begin value
+    Expr calc_load(const Expr &e) {
+        if (e.as<Variable>()) {
+            return calc_load(get_val(e));
+        }
+        if (const Cast *maybe_cast = e.as<Cast>()) {
+            return calc_load(maybe_cast->value);
+        }
+        if (const Shuffle *maybe_shuffle = e.as<Shuffle>()) {
+            if (maybe_shuffle->is_slice()) {
+                Expr maybe_load = calc_load(maybe_shuffle->vectors[0]);
+                if (!maybe_load.defined()) {
+                    return nullptr;
+                }
+                const Load *res = maybe_load.as<Load>();
+                const Ramp *ramp = get_val(res->index).as<Ramp>();
+                if (!ramp) {
+                    return nullptr;
+                }
+                Expr shifted_ramp = Ramp::make(ramp->base + maybe_shuffle->slice_begin(),
+                                                ramp->stride, ramp->lanes);
+                Expr shifted_load = Load::make(res->type, res->name, shifted_ramp,
+                                                res->image, res->param, res->predicate);
+                return shifted_load;
+            }
+            if (maybe_shuffle->is_concat()) {
+                return are_continuous_vectors(maybe_shuffle->vectors);
+            }
+        }
+        if (const Load *maybe_load = e.as<Load>()) {
+            return maybe_load;
+        }
+        return nullptr;
+    }
+
+    // Vtmpy helps in sliding window ops of the form a*v0 + b*v1 + v2.
```

Comment:
I agree, front-end intrinsic doesn't seem like a viable option to me also.
Also the complexity in the code is due to trying different permutations of the expressions collected in mpys vector. The vrmpy doesn't test any combination and marks this as future work in a TODO comment. Even the shuffle in the vrmpy doesn't contribute in reducing the time complexity.